internal/controller/controllers/clusterdeployments_controller.go

```diff
@@ -103,6 +106,38 @@ func (r *ClusterDeploymentsReconciler) Reconcile(ctx context.Context, req ctrl.R
 		return ctrl.Result{}, nil
 	}
 
+	if clusterDeployment.ObjectMeta.DeletionTimestamp.IsZero() { // clusterDeployment no being deleted
```

Comment:
Once we add an owner reference we can probably look at only agent cluster install no?

Reply:
I guess the question we should ask: when do we want to have that cleanup triggered? on `clusterDeployment` deletion or `agentClusterInstall` deletion (which may happen either with that resource deletion or when we delete `clusterDeployment` because of the ownerRef)

Reply:
Actually infra env reference to cluster deployment and not to agent cluster install so we don't really have a way to enforce it.

Reply:
I took a second look at the code that was added with `AgentClusterInstall`.
We stop the reconcile in case there is no `AgentClusterInstall` to be found for the `clusterDeployment`, meaning we will not call `RegisterClusterInternal()`

https://github.com/openshift/assisted-service/blob/d3657fd59d41d5a9301bf96d2adea27974763a8d/internal/controller/controllers/clusterdeployments_controller.go#L106-L110

So maybe it does make sense to deregister (and delete agents) when `AgentClusterInstall` gets removed? seems like a more "symmetrical". what do you think?

BTW, not sure how that'll play out with the day2 flow, I need to look at that as well.

Reply:
I would trigger cleanup on AgentClusterInstall rather than ClusterDeployment.

Reply:
@dgoodwin for what reason?

Reply:
Teardown logic is intricately linked to the provision logic, so it makes the most sense for the finalizers to be tied to implementation specific resource. I am also hoping cluster install controllers *never* write to the ClusterDeployment, and hive controllers never write to an external cluster install CR.

Reply:
We reconcile on cluster deployment as the main resource, so once it was deleted and we need to do the cleanup the reconcile loop will not find cluster deployment so how should it work?

Reply:
Hrm, if your AgentClusterInstall has a finalizer, and an OwnerRef to ClusterDeployment, will the ClusterDeployment disappear before your AgentClusterInstall? I am thinking it will remain.